src/reducers.js

```diff
@@ -0,0 +1,182 @@
+   @method _nextCellType
+   @param {String} type Cell type.
+   @return {String}
+*/
+function _nextCellType(type) {
+    const index = interactiveCellTypes.indexOf(type);
+
+    if (index < 0) {
+        return 'empty';
+    }
+
+    return interactiveCellTypes[(index + 1) % interactiveCellTypes.length];
+}
+
+/**
+   Game state reducer.
+*/
+function game(state = predefinedGames[INITIAL_GAME_INDEX], action) {
+
+    // Game index and reset.  Return a predefined game.
+    if (action.type === CHANGE_GAME_INDEX || action.type === RESET_GAME) {
+        return predefinedGames[action.index];
+    }
+
+    // If its any other action we don't know about, just return state back.
+    if (action.type !== CLICK_CELL) {
+        return state;
+    }
+
+    // Out of bounds error. Just return same board.
+    if (action.row < 1 ||
+        action.row >= state.board.length ||
+        action.column < 1 ||
+        action.column >= state.board[0].length ||
+        state.board[action.row][action.column].type === 'coral') {
+        return state;
+    }
+
+    // Create new immutable board.
+    const board = state.board.map(row => {
+        return row.map(cell => Object.assign({}, cell));
+    });
+
+    // Update cell type based on what was clicked.
+    const cell = board[action.row][action.column];
+    cell.type = _nextCellType(cell.type);
+
+    const rowConstraintCell = board[action.row][0];
+    const colConstraintCell = board[0][action.column];
+
+    // Update row constraints state.
+    const row = board[action.row];
+    const column = board.map(row => row[action.column]);
+
+    if (!row.some(cell => cell.type === 'empty')) {
+        let numClownfishs = 0;
+
+        row.forEach(cell => {
+            if (cell.type === 'clownfish') {
+                numClownfishs++;
+            }
+        });
+
+        rowConstraintCell.fulfilled = rowConstraintCell.value === numClownfishs;
+        rowConstraintCell.unfulfilled = !rowConstraintCell.fulfilled;
+    }
+    else {
+        rowConstraintCell.fulfilled = rowConstraintCell.unfulfilled = false;
+    }
+
+    // Update column constraints state.
+    if (!column.some(cell => cell.type === 'empty')) {
+        let numClownfish = 0;
+
+        column.forEach(cell => {
+            if (cell.type === 'clownfish') {
+                numClownfish++;
+            }
+        });
+
+        colConstraintCell.fulfilled = colConstraintCell.value === numClownfish;
+        colConstraintCell.unfulfilled = !colConstraintCell.fulfilled;
+    }
+    else {
+        colConstraintCell.fulfilled = colConstraintCell.unfulfilled = false;
+    }
+
+    // Pull all row and column constraints. We'll use to determine if board is showing a winner.
+    const cols = board[0];
+    const rows = board.map(row => row[0]);
+
+    // Return a new version of the game state.
+    return {
+        title: state.title,
+        board,
+        winner: rows.every(cell => cell.fulfilled) && cols.every(cell => cell.fulfilled),
```

Comment:
So I'm a little on the fence about this, but I actually don't think I'd include this `winner` state here, since it can be entirely derived from your other state. Typically I try to normalize my redux state as much as possible and derive what I need during the `render` function of some component. 

The reason I'm iffy on this is that in a more complicated game you might actually want to transition to some entirely new state when there's a winner. But since, in this case, you actually let the user continue to interact, I'd pull this `winner` variable out of state and derive it in `render`.